### ISSUE TITLE
ban_record states that banned player doesn't exist

### ISSUE THREAD

**Issue by rubenwardy** [shivajiva101/sban#33]
> /ban_record LoneWolfHT
> Unknown player!

sban.sqlite:

```
sqlite> SELECT * FROM name WHERE name="LoneWolfHT";
-- 0 records --
sqlite> SELECT * FROM active WHERE name="LoneWolfHT";
6170|LoneWolfHT|Lone_Wolf|1599147415|Impersonating staff|0|(13158,2,-81)
```

Nothing in the stdout logs

Mods are capture the flag + MT-CTF/rules, MT-CTF/servermods, sauth, sban

**Comment by shivajiva101** [comment]
Interesting and thanks for the information

**Comment by shivajiva101** [comment]
Any more updates on this occurring? I suspect it happened in data migration unless you can tell me otherwise.
Does id 6170 in the name table have an entry?

**Comment by LoneWolfHT** [comment]
This happens a lot, I forget specific examples though. I will post here with extra info if the issue pops up again.

**Comment by shivajiva101** [comment]
When a player joins the server the db is queried for an id that matches the name or the ip address, if the ip address is known but the name is not, then a new entry is created in the name table under that id. If the name is known but not the ip address, then a new entry is created in the address table under that id. If neither is known a new id is created and name and ip address are stored under that id in the respective tables.

Considering the fact you can't ban a player who doesn't have an id, unless you have server privilege, and the command requires a player name before it creates an id to insert the given data for the ban, I can't see a way for this to happen unless there is a transaction failure with the sqlite instance. 

Even considering that possibility you would be aware that was the circumstances of how it originated, presumably you had logged into the server many times prior to trying ban yourself, or having another moderator execute the ban command on your account? I'm just pointing out that the mod had multiple opportunities to actually create a name entry for that nick, even if your ip address was different every single time, and yet apparently failed every time...seems highly unlikely odds of that occurring. I really need more information about when this occurs as it makes no sense whatsoever.

I can code some checks in and log them to see if it can be pinned down I guess, as a proactive step to getting some insight on this

@LoneWolfHT
update: data integrity check added on mod load to log in the debug text file, I'd appreciate you running the latest to see if you can get me some more info on the frequency it happens and what method was used when banning them

example of output in log:
```
2021-12-11 17:11:30: ACTION[Main]: [sban] Data integrity check...
2021-12-11 17:11:30: ACTION[Main]: [sban] id: 255897 foofy test Tue 27 Aug 2019 22:40:00 BST test ban is orphaned!

```
inserted with sqlite for test purposes

**Comment by shivajiva101** [comment]
@LoneWolfHT Possible fix for this in latest version. I think joining with a different nick and banning it could have left the db with an orphaned ban record as sban was failing to write new names for a known id since introducing the temp join cache to save a couple of function calls. You should update asap

**Comment by LoneWolfHT** [comment]
> @LoneWolfHT Possible fix for this in latest version. I think joining with a different nick and banning it could have left the db with an orphaned ban record as sban was failing to write new names for a known id since introducing the temp join cache to save a couple of function calls. You should update asap

Sauth seemed to update fine, but sban gave me this: 
```
minetest@ctfserver:~/minetest/worlds/ctf$ sqlite3 sban.sqlite
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> .read sban_update.sql
Error: near line 149: no such table: bans
Error: near line 166: no such table: bans
Error: near line 167: no such table: bans
Error: near line 173: no such table: bans
Error: near line 187: no such table: bans
Error: near line 201: UNIQUE constraint failed: config.setting
Error: near line 202: UNIQUE constraint failed: config.setting
Error: near line 205: no such table: playerdata
Error: near line 207: no such table: playerdata
Error: near line 229: no such table: bans
Error: near line 232: no such table: players
Error: near line 233: no such table: playerdata
Error: near line 234: no such table: version
sqlite> .exit```

**Comment by shivajiva101** [comment]
looks like you ran the update on an already updated database! It's only required if the server doesn't start and complains the db needs updating, perhaps I should rephrase the instructions to make them a little clearer

**Comment by LoneWolfHT** [comment]
Just found a player who's been around for about 2 days who doesn't show up in /ban_record; Neither of the commands used in the OP (on sban.sqlite) work.

Nothing in the logs but this, which is unrelated:
```
2021-12-16 10:22:51: WARNING[Server]: Call to deprecated function 'getpos', please use 'get_pos' at ...est/bin/../worlds/ctf/worldmods/servermods/sban/init.lua:822
```

**Comment by shivajiva101** [comment]
Interesting, anything unusual about their nick or ip addresses? Can you validate they showed up after the update in your log? Did they join on an already known IP? 
Some set of conditions had to be met, were they banned in the first session they played on the server? Could really do with you having a root around the database to see if you can notice anything unusual. db_exec logs sql errors to the log and to the chat console so a failure to write the player record would have to show up, are there any traces of this player in the db?

meanwhile I'll fix the warning caused by line 822 :+1: 

I noticed from the line number you are a couple of commits earlier than latest, shouldn't make any difference though and I'm not sure what you mean by this `Neither of the commands used in the OP (on sban.sqlite) work.` could you rephrase that sentence or explain in a little more detail pls

**Comment by LoneWolfHT** [comment]
I just discovered their last join was before the update by a few hours, sorry for the (second) false alarm lol
I have figured out how to investigate any future problems more closely at least

**Comment by shivajiva101** [comment]
I believe commit https://github.com/shivajiva101/sban/commit/7ce051fb77b228528c9580d1eed2906e02b1a41c fixes this issue so I'll close this for now. Feel free to reopen the issue if the problem resurfaces with the latest code
